app/validators/between_validator.rb

```diff
@@ -0,0 +1,10 @@
+class BetweenValidator < ActiveModel::EachValidator
```

Comment:
Unfortunately, the comparison validator doesn't work the way we want it to. We could also use the numericality validator with the `less_than_or_equal_to: :max_hours` option.

The problem is the localisation and that the validator attempts each option, one by one, then raises the first error found.

e.g.

```ruby
validates :hours_completed, numericality: { greater_than_or_equal_to: 1, less_than_or_equal_to: :max_hours, only_integer: true }
```

- Check if integer, if not, raise error.
- Check if greater than or equal to 1, raise error.
  - This error message is then interpolated with `count`, the value passed into the option. In this case, `1`.
- Check if less than or equal to :max_hours method return value, raise error.
  - This error message is then interpolated with `count`, the value passed into the option. In this case, the return value of `:max_hours`.

See more: https://guides.rubyonrails.org/i18n.html#error-message-interpolation

So, in our case, the `less_than_or_equal_to` will be fine, because our lower bound is static, `1`. The upper bound is available in the i18n message as `%{count}`. We can do: `less_than_or_equal_to: Enter the number of hours between 1 and %{count}`.

But for `greater_than_or_equal_to`, we only have the lower bound available as `%{count}` and we cannot determine the upper bound for the message.

## Alternative 2

```ruby
validates :hours_completed, numericality: { in: :my_range }

def my_range
  1..max_hours
end
```

This also works, but it has a similar issue. The `%{count}` value passed to the message interpolation is the range itself. The problem is that it's _already_ interpolated, therefore, `#to_s` is implicitly called on it, and it looks like: `is not in 1..20`.

We also can't do things like `%{count.first}` and `%{count.last}`. :/

This feels like an area Rails could be improved though. A PR would take far too long to get into the next release of Rails though :P

Reply:
Oh, that makes sense. Leave it like this no worries 👍🏻